src/hash_http_content/hasher.py

```diff
@@ -0,0 +1,219 @@
+
+    def _handle_raw_bytes(self, contents: bytes, encoding: str) -> HandlerResult:
+        """Handle bytes in an unspecified format or encoding."""
+        logging.debug("Handling content as raw bytes")
+        digest: str = get_hash_digest(self._hash_algorithm, contents)
+        return HandlerResult(digest, contents)
+
+    def _handle_plaintext(self, contents: bytes, encoding: str) -> HandlerResult:
+        """Handle plaintext contents."""
+        logging.debug("Handling content as plaintext")
+        if encoding:
+            contents = bytes(contents.decode(encoding), self._default_encoding)
+        digest: str = get_hash_digest(self._hash_algorithm, contents)
+        return HandlerResult(digest, contents)
+
+    def _handle_json(self, contents: bytes, encoding: str) -> HandlerResult:
+        """Handle JSON contents."""
+        logging.debug("Handling content as JSON")
+        # Translate the original encoding to utf-8
+        if encoding:
+            json_str = str(contents, encoding)
+        else:
+            json_str = str(contents, self._default_encoding)
+
+        json_data = json.loads(json_str)
+        # Sort the keys to make this deterministic
+        json_bytes = bytes(
+            json.dumps(json_data, separators=(",", ":"), sort_keys=True),
+            self._default_encoding,
+        )
+
+        digest: str = get_hash_digest(self._hash_algorithm, json_bytes)
+
+        return HandlerResult(digest, json_bytes)
+
+    def _handle_html(self, contents: bytes, encoding: str) -> HandlerResult:
+        """Handle an HTML page."""
+        logging.debug("Handling content as HTML")
+        self.__init_browser()
+
+        if encoding:
+            html = str(contents, encoding)
+        else:
+            html = str(contents, self._default_encoding)
+
+        logging.debug("Setting page contents and rendering")
+        page: Page = asyncio.get_event_loop().run_until_complete(
+            self._browser.newPage()
+        )
+        asyncio.get_event_loop().run_until_complete(page.setContent(html))
+        page_contents: str = asyncio.get_event_loop().run_until_complete(page.content())
+        asyncio.get_event_loop().run_until_complete(page.close())
+
+        logging.debug("Parsing rendered page contents")
+        soup: BeautifulSoup = BeautifulSoup(page_contents, "lxml")
+        text_elements = soup.find_all(text=True)
+        visible_text_elements = filter(self._is_visible_element, text_elements)
+        visible_text = " ".join(t.strip() for t in visible_text_elements if t.strip())
+        visible_bytes = bytes(visible_text, self._default_encoding)
+
+        digest: str = get_hash_digest(self._hash_algorithm, visible_bytes)
+
+        return HandlerResult(digest, visible_bytes)
+
+    def hash_url(self, url: str) -> UrlResult:
+        """Get a hash of the contents of the provided URL."""
+        logging.debug("Hashing provided URL '%s'", url)
+        redirect_status_codes = [301, 307, 308]
```

Comment:
I'm no expert on the intricacies of redirects, but there are other redirect codes besides these three- is there are reason you are excluding them (e.g. 302, 303, 305)?  https://tools.ietf.org/html/rfc7231#section-6.4

Reply:
I overlooked adding `302` when I was setting up that list, but my reasoning was to only count one type of redirect (per rfc7231):

> 1.  Redirects that indicate the resource might be available at a
       different URI, as provided by the Location field, as in the
       status codes 301 (Moved Permanently), 302 (Found), and 307
       (Temporary Redirect).

This is in line with the creation of `308` as described in https://tools.ietf.org/html/rfc7238.

Reply:
Excellent question and sharp eyes @dav3r!

@mcdonnnj, you should add the `302` code and add a comment explaining why you are only counting that one type of redirect.  That isn't immediately clear to me.

Reply:
I have added `302` and explanatory comments in a988651. Please let me know if there are any clarifications or further information you think it needs.

Reply:
I'm good with your explanation in https://github.com/cisagov/hash-http-content/commit/a988651494d4639adc2d583a2d3095cc1eec369a - thanks!